### PR TITLE
Reconfigure codecov to reduce false alarm rate.

### PR DESCRIPTION
**Pull request by twiecki**

### CODE REVIEW

File: codecov.yml [hunk 1]
```diff
@@ -1,18 +1,28 @@
...
-
+        threshold: 1%
+        base: auto
     patch:
       default:
-        target: 100%
+        # basic
+        target: 50%
+        threshold: 1%
+        base: auto
```
Comment (brandonwillard) [inline codecov.yml:20]: We need 100% coverage for newly added code.  While this may be annoying for some forms of refactoring, even in that context it's at least prompting us to check and make sure we didn't actually reduce coverage while refactoring.

File: codecov.yml [hunk 2]
```diff
@@ -1,18 +1,28 @@
 codecov:
-  require_ci_to_pass: true
-
-comment:
-    behavior: default
-    branches:
-        - "master"
+  require_ci_to_pass: no
```
Comment (brandonwillard) [inline codecov.yml:2]: Why this change?

**Review by brandonwillard — CHANGES_REQUESTED** [review]
Is this PR for #175?  That issue addresses the false alarms caused by partial reports (i.e. unfinished test workflows), but the patch failures reported after all the tests have run are _not_ really false alarms in the same sense.

Again, the patch failures may be somewhat annoying during refactoring, but we need to look at the coverage changes manually and determine whether or not they're due to refactoring.  Without the reported failures in those cases, we would constantly be given the wrong impression and end up merging more untested code more often.  Given the current state of this project, we can't afford it.

**Review by brandonwillard — COMMENTED** [review]
I think @OriolAbril might've mentioned that this option would help: [`notify.after_n_builds`](https://docs.codecov.io/docs/codecovyml-reference#codecovnotifyafter_n_builds).

**Review by brandonwillard — COMMENTED** [review]
The build process should finish even though the tests didn't run, but for some reason it's stuck waiting for a response from Codecov.  That might be something we need to fix on the repo and/or workflow side, though.

**Comment by twiecki** [comment]
Should we merge?

**Comment by brandonwillard** [comment]
Let's try pushing a temporary code change so that it will trigger the tests, then we can see if the intermediate failures issue is still a problem and whether or not we need to use something like `notify.after_n_builds`.

**Comment by MarcoGorelli** [comment]
> I think @OriolAbril might've mentioned that this option would help: [`notify.after_n_builds`](https://docs.codecov.io/docs/codecovyml-reference#codecovnotifyafter_n_builds).

just tried that out in https://github.com/pymc-devs/pymc3/pull/4255 - some jobs have finished, and so far there's been no prematurely reported failure

**Comment by brandonwillard** [comment]
I just pushed a commit that attempts to send coverage information only once all the matrix jobs have finished.  @dfm, set up a convenient `combine` job that serves as a single indicator for a successful run of all the matrix-generated tests.  Its primary purpose was to serve as a single GitHub required check (i.e. instead of using an arguably variable number of previously generated matrix tests), but I think it can serve as a single point for uploading the cumulative coverage results, as well.

I'm not sure if the exact change in that commit is the correct one, but the general approach seems like a viable way to get what we're looking for without introducing new parameters.

**Comment by brandonwillard** [comment]
OK, this approach seems to work.  It also has the nice side effect of producing an artifact for the all the raw coverage data (e.g. to be used for something like #107).

To summarize: this configuration is using Action's artifacts to download and append coverage information to a single `coverage.xml` across each test produced by the matrix job, then a final job is run that pushes the coverage file to Codecov.
This prevents any premature failure reports and is robust to workflow/matrix changes (i.e. we won't have to change parameters to match the number of matrix runs).
